packages/libs/error-stack/src/frame/frame_impl.rs

```diff
@@ -0,0 +1,216 @@
+    fn source_mut(&mut self) -> Option<&mut Frame>;
+
+    /// Provide values which can then be requested.
+    #[cfg(nightly)]
+    fn provide<'a>(&'a self, demand: &mut Demand<'a>);
+}
+
+#[repr(C)]
+struct ContextFrame<C> {
+    context: C,
+    location: &'static Location<'static>,
+    source: Option<Box<Frame>>,
+}
+
+// SAFETY: `type_id` returns `C` and `C` is the first field in `#[repr(C)]`
+unsafe impl<C: Context> FrameImpl for ContextFrame<C> {
+    fn kind(&self) -> FrameKind<'_> {
+        FrameKind::Context(&self.context)
+    }
+
+    fn type_id(&self) -> TypeId {
+        TypeId::of::<C>()
+    }
+
+    fn location(&self) -> &'static Location<'static> {
+        self.location
+    }
+
+    fn source(&self) -> Option<&Frame> {
+        self.source.as_ref().map(Box::as_ref)
+    }
+
+    fn source_mut(&mut self) -> Option<&mut Frame> {
+        self.source.as_mut().map(Box::as_mut)
+    }
+
+    #[cfg(nightly)]
+    fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
+        Context::provide(&self.context, demand);
+    }
+}
+
+#[repr(C)]
+struct AttachmentFrame<A> {
+    attachment: A,
+    location: &'static Location<'static>,
+    source: Option<Box<Frame>>,
+}
+
+// SAFETY: `type_id` returns `A` and `A` is the first field in `#[repr(C)]`
+unsafe impl<A: 'static + Send + Sync> FrameImpl for AttachmentFrame<A> {
+    fn kind(&self) -> FrameKind<'_> {
+        FrameKind::Attachment(AttachmentKind::Opaque(&self.attachment))
+    }
+
+    fn type_id(&self) -> TypeId {
+        TypeId::of::<A>()
+    }
+
+    fn location(&self) -> &'static Location<'static> {
+        self.location
+    }
+
+    fn source(&self) -> Option<&Frame> {
+        self.source.as_ref().map(Box::as_ref)
+    }
+
+    fn source_mut(&mut self) -> Option<&mut Frame> {
+        self.source.as_mut().map(Box::as_mut)
+    }
+
+    #[cfg(nightly)]
+    fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
+        demand.provide_ref(&self.attachment);
+    }
+}
+
+#[repr(C)]
+struct PrintableAttachmentFrame<A> {
+    attachment: A,
+    location: &'static Location<'static>,
+    source: Option<Box<Frame>>,
+}
+
+// SAFETY: `type_id` returns `A` and `A` is the first field in `#[repr(C)]`
+unsafe impl<A: 'static + Debug + Display + Send + Sync> FrameImpl for PrintableAttachmentFrame<A> {
+    fn kind(&self) -> FrameKind<'_> {
+        FrameKind::Attachment(AttachmentKind::Printable(&self.attachment))
+    }
+
+    fn type_id(&self) -> TypeId {
+        TypeId::of::<A>()
+    }
+
+    fn location(&self) -> &'static Location<'static> {
+        self.location
+    }
+
+    fn source(&self) -> Option<&Frame> {
+        self.source.as_ref().map(Box::as_ref)
+    }
+
+    fn source_mut(&mut self) -> Option<&mut Frame> {
+        self.source.as_mut().map(Box::as_mut)
+    }
+
+    #[cfg(nightly)]
+    fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
+        demand.provide_ref(&self.attachment);
+    }
+}
+
+impl Frame {
+    /// Crates a frame from a [`Context`].
```

Comment:
```suggestion
    /// Creates a frame from a [`Context`].
```

Reply:
Is it actually from a [`Context`]?

Reply:
Yeah I see the trait bound now, it was hidden by GH. I asked because previously we had weird behaviour with the compat stuff, etc. So I was wondering if it was a plain generic, I see it's not :)